src/main/java/com/o19s/es/ltr/query/RankerQuery.java

```diff
@@ -0,0 +1,320 @@
+            }
+            float modelScore = ranker.score(d);
+            return Explanation.match(modelScore, " LtrModel: " + ranker.name() + " using features:", subs);
+        }
+
+        @Override
+        public float getValueForNormalization() throws IOException {
+            // XXX: may produce inconsistent scores if featured queries
+            // are run individually and a Similarity that implements queryNorm is used.
+            // queryNorm will disappear in lucene 7.
+            // Should we hack something or warn if this query is being used with ClassicSimilarity?
+            float sum = 0.0f;
+            for (Weight w : weights) {
+                sum += w.getValueForNormalization();
+            }
+            return sum ;
+        }
+
+        @Override
+        public void normalize(float norm, float boost) {
+            for(Weight w : weights) {
+                w.normalize(norm, boost);
+            }
+        }
+
+        @Override
+        public RankerScorer scorer(LeafReaderContext context) throws IOException {
+            ArrayList<Scorer> scorers = new ArrayList<>(weights.length);
+            List<DocIdSetIterator> subIterators = new ArrayList<>(weights.length);
+            for (int i = 0; i < weights.length; i++) {
+                Scorer scorer = weights[i].scorer(context);
+                if (scorer == null) {
+                    scorer = new NoopScorer(this, context.reader().maxDoc());
+                }
+                scorers.add(scorer);
+                subIterators.add(scorer.iterator());
+            }
+            DocIdSetIterator rankerIterator = new NaiveDisjunctionDISI(DocIdSetIterator.all(context.reader().maxDoc()), subIterators);
+            return new RankerScorer(scorers, rankerIterator);
+        }
+
+        class RankerScorer extends Scorer {
+            /**
+             * NOTE: Switch to ChildScorer and {@link #getChildren()} if it appears
+             * to be useful for logging
+             */
+            private final ArrayList<Scorer> scorers;
+            private final DocIdSetIterator iterator;
+            private final LtrRanker.DataPoint dataPoint;
+
+            RankerScorer(ArrayList<Scorer> scorers, DocIdSetIterator iterator) {
+                super(RankerWeight.this);
+                this.scorers = scorers;
+                this.iterator = iterator;
+                dataPoint = ranker.newDataPoint();
+            }
+
+            @Override
+            public int docID() {
+                return iterator.docID();
+            }
+
+            @Override
+            public float score() throws IOException {
+                for (int i = 0; i < scorers.size(); i++) {
+                    Scorer scorer = scorers.get(i);
+                    if (scorer.docID() == docID()) {
+                        dataPoint.setFeatureScore(i, scorer.score());
+                    } else {
+                        dataPoint.setFeatureScore(i, 0);
```

Comment:
Also I'm struggling to visualize how a sparse lucene query would look like, today it's clear that we make the bold assumption that the model is dense:
`initialFeatures.length == queries.length == scorers.length == featureVector.length >= numberOfDistinctFeatureUsedInTheModel`
and that we can use the array index as a feature identifier.
If we go with a sparse approach this equality will break somewhere and we will  need some indirection to infer the featureIdentifier.

Looking at the solr implementation I see that it's relatively dense as well https://github.com/apache/lucene-solr/blob/master/solr/contrib/ltr/src/java/org/apache/solr/ltr/LTRScoringQuery.java#L406 despite having a strong distinction between Sparse and Dense scorers. For solr the feature remains identified by its index in the array.

I agree that we currently lack a way to describe missing values.

If dense vs sparse is just something to allow identifying missing values then it should be relatively easy to do.
On the other hand if sparse means that the equality I wrote earlier is not verified then I think we need more examples to clarify its usage.